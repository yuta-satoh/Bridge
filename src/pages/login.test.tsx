import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import Login from './login';

jest.mock('next/router', () => ({
    useRouter() {
      return {
        replace: jest.fn(),
      };
    },
}));

const fetchMockOk = () => {
    return new Promise((resolve) => {
        resolve({
            ok: true,
            status: 200,
            statusText: 'OK',
        })
    })
}

const fetchMockNg = () => {
    return new Promise((resolve) => {
        resolve({
            ok: false,
            status: 401,
            statusText: 'Failed',
        })
    })
}

describe('ログインページ', () => {
    test('ログインページが正常に描画されました', () => {
        const { asFragment } = render(<Login />);
        expect(asFragment()).toMatchSnapshot()
    })

    test('正常なメールアドレスとパスワードを入力してログイン', async () => {
        global.fetch = jest.fn().mockImplementation(fetchMockOk);
        // setupがおすすめされてるので先に実行
        const user = userEvent.setup()
        // Login.tsxをレンダリング
        const { asFragment } = render(<Login />);

        // メールアドレスを入力して確認
        const inputEmail = screen.getByLabelText('メールアドレス');
        await user.type(inputEmail, 'rakus@example.com');
        expect(inputEmail).toHaveValue('rakus@example.com');
        // パスワードを入力して確認
        const inputPassword = screen.getByLabelText('パスワード')
        await user.type(inputPassword, 'rakus');
        expect(inputPassword).toHaveValue('rakus');

        // ログインボタンをクリック
        await user.click(screen.getByRole('button', { name: 'ログイン →' }))

        expect(asFragment()).toMatchSnapshot()
    })

    test('異常なメールアドレスとパスワードを入力してログイン', async () => {
        global.fetch = jest.fn().mockImplementation(fetchMockNg);
        // setupがおすすめされてるので先に実行
        const user = userEvent.setup()
        // Login.tsxをレンダリング
        const { asFragment } = render(<Login />);

        // メールアドレスを入力して確認
        const inputEmail = screen.getByLabelText('メールアドレス');
        await user.type(inputEmail, 'notUser@example.com');
        expect(inputEmail).toHaveValue('notUser@example.com');
        // パスワードを入力して確認
        const inputPassword = screen.getByLabelText('パスワード')
        await user.type(inputPassword, 'notUser');
        expect(inputPassword).toHaveValue('notUser');

        // ログインボタンをクリック
        await user.click(screen.getByRole('button', { name: 'ログイン →' }))
        
        // エラーメッセージが表示されてるとOK
        expect(screen.getByText('メールアドレスかパスワードが間違っています').textContent).toBe('メールアドレスかパスワードが間違っています')

        expect(asFragment()).toMatchSnapshot()
    })
})
