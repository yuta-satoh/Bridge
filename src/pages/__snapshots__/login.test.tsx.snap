// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ログインページ ログインページが正常に描画されました 1`] = `
<DocumentFragment>
  <main
    class="main"
  >
    <div
      class="body"
    >
      <div
        class="title"
      >
        <h1
          class="h1"
        >
          LOGIN
        </h1>
        <p
          class="p"
        >
          ログイン
        </p>
      </div>
      <form
        class="form"
      >
        <div
          class="inputItems"
        >
          <label
            class="label"
            for="email"
          >
            メールアドレス
          </label>
          <br />
          <input
            class="inputParts border border-neutral-500 rounded pl-2.5"
            id="email"
            name="email"
            placeholder="例：bridge@example.com"
            type="email"
          />
          <p
            class="nope"
          >
            ※メールアドレスを入力して下さい
          </p>
        </div>
        <div
          class="inputItems"
        >
          <label
            class="label"
            for="password"
          >
            パスワード
          </label>
          <br />
          <input
            class="inputParts border border-neutral-500 rounded pl-2.5"
            id="password"
            name="password"
            placeholder="例：abcdef123456"
            type="password"
          />
          <p
            class="nope"
          >
            ※パスワードを入力して下さい
          </p>
        </div>
        <div
          class="buttonArea"
        >
          <p
            class="error"
          />
          <button
            class="loginButton"
            type="submit"
          >
            ログイン
            <span
              class="buttonSpan"
            >
              →
            </span>
          </button>
        </div>
      </form>
      <div
        class="title"
      >
        <h2
          class="subTitle"
        >
          はじめてご利用のお客様はこちら
        </h2>
      </div>
      <div
        class="loginLink"
      >
        <a
          href="/userRegister"
        >
          <button
            class="linkButton"
            type="button"
          >
            新規会員登録
            <span
              class="buttonSpan"
            >
              →
            </span>
          </button>
        </a>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`ログインページ 正常なメールアドレスとパスワードを入力してログイン 1`] = `
<DocumentFragment>
  <main
    class="main"
  >
    <div
      class="body"
    >
      <div
        class="title"
      >
        <h1
          class="h1"
        >
          LOGIN
        </h1>
        <p
          class="p"
        >
          ログイン
        </p>
      </div>
      <form
        class="form"
      >
        <div
          class="inputItems"
        >
          <label
            class="label"
            for="email"
          >
            メールアドレス
          </label>
          <br />
          <input
            class="inputParts border border-neutral-500 rounded pl-2.5"
            id="email"
            name="email"
            placeholder="例：bridge@example.com"
            type="email"
          />
          <p
            class="nope"
          >
            ※メールアドレスを入力して下さい
          </p>
        </div>
        <div
          class="inputItems"
        >
          <label
            class="label"
            for="password"
          >
            パスワード
          </label>
          <br />
          <input
            class="inputParts border border-neutral-500 rounded pl-2.5"
            id="password"
            name="password"
            placeholder="例：abcdef123456"
            type="password"
          />
          <p
            class="nope"
          >
            ※パスワードを入力して下さい
          </p>
        </div>
        <div
          class="buttonArea"
        >
          <p
            class="error"
          />
          <button
            class="loginButton"
            type="submit"
          >
            ログイン
            <span
              class="buttonSpan"
            >
              →
            </span>
          </button>
        </div>
      </form>
      <div
        class="title"
      >
        <h2
          class="subTitle"
        >
          はじめてご利用のお客様はこちら
        </h2>
      </div>
      <div
        class="loginLink"
      >
        <a
          href="/userRegister"
        >
          <button
            class="linkButton"
            type="button"
          >
            新規会員登録
            <span
              class="buttonSpan"
            >
              →
            </span>
          </button>
        </a>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`ログインページ 異常なメールアドレスとパスワードを入力してログイン 1`] = `
<DocumentFragment>
  <main
    class="main"
  >
    <div
      class="body"
    >
      <div
        class="title"
      >
        <h1
          class="h1"
        >
          LOGIN
        </h1>
        <p
          class="p"
        >
          ログイン
        </p>
      </div>
      <form
        class="form"
      >
        <div
          class="inputItems"
        >
          <label
            class="label"
            for="email"
          >
            メールアドレス
          </label>
          <br />
          <input
            class="inputParts border border-neutral-500 rounded pl-2.5"
            id="email"
            name="email"
            placeholder="例：bridge@example.com"
            type="email"
          />
          <p
            class="nope"
          >
            ※メールアドレスを入力して下さい
          </p>
        </div>
        <div
          class="inputItems"
        >
          <label
            class="label"
            for="password"
          >
            パスワード
          </label>
          <br />
          <input
            class="inputParts border border-neutral-500 rounded pl-2.5"
            id="password"
            name="password"
            placeholder="例：abcdef123456"
            type="password"
          />
          <p
            class="nope"
          >
            ※パスワードを入力して下さい
          </p>
        </div>
        <div
          class="buttonArea"
        >
          <p
            class="error"
          >
            メールアドレスかパスワードが間違っています
          </p>
          <button
            class="loginButton"
            type="submit"
          >
            ログイン
            <span
              class="buttonSpan"
            >
              →
            </span>
          </button>
        </div>
      </form>
      <div
        class="title"
      >
        <h2
          class="subTitle"
        >
          はじめてご利用のお客様はこちら
        </h2>
      </div>
      <div
        class="loginLink"
      >
        <a
          href="/userRegister"
        >
          <button
            class="linkButton"
            type="button"
          >
            新規会員登録
            <span
              class="buttonSpan"
            >
              →
            </span>
          </button>
        </a>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
